### PR TITLE
plugins: refactor initialization

### LINKED ISSUE #315: Don't use flask_babel at import time
At import time the language domain is unknown, using `flask_babel` to translate values at import time makes no sense . . see discussion at https://github.com/searxng/searxng/pull/308#discussion_r706385553

This issue is related to a common design mistake: [Don’t run code at import time](https://www.benkuhn.net/importtime). Here the problem is, that some modules try to initialize their values which are depend on an application context. Such values should be initialized within the application (means: by the application).

### PR DESCRIPTION
**Pull request by dalf**
## What does this PR do?

Add a new function "init" call when the app starts.
The function can:
* return False to disable the plugin (case of ahmia_filter plugin).
* modify the Flask app.

See https://github.com/dalf/searxng/blob/mod-plugins/searx/plugins/ahmia_filter.py

Reference to the plugins are by Python module names instead of plugin names (related to https://github.com/searxng/searxng/issues/315)

The function `searx.plugins.load_plugin` looks like `searx.engines.load_engines`.

BREAKING CHANGE: the user preferences (cookies) about the plugins will be lost.

Embedded plugins are automatically detected.

## Why is this change important?

With this change a plugin can change the web application:
* add / remove / replace some HTTP endpoints
* add middleware, etc...

Draft:
* ~if would be nice if the `init` could replace the application: `app = Flask_Babel(app)`.~
* ~perhaps implement a temporary workaround to avoid a breaking change about the cookies.~ (it just requires to go the preferences and make some changes and save)

## How to test this PR locally?

* `make run`
* check it works with an external plugin including static resources

## Related issues

Close #315

### CODE REVIEW

File: searx/webapp.py [hunk 1]
```diff
@@ -1351,6 +1332,27 @@ def page_not_found(_e):
...
+#  __name__ != "__main__" if searx.webapp is imported (make test, make docs, uwsgi...)
+# see run() at the end of this file : searx_debug activates the reload feature.
+werkzeug_reloader = flask_run_development or (searx_debug and __name__ == "__main__")
+
+# initialize the engines except on the first run of the werkzeug server.
+if (not werkzeug_reloader
+    or (werkzeug_reloader
+        and os.environ.get("WERKZEUG_RUN_MAIN") == "true") ):
+    plugin_initialize(app)
+    search_initialize(enable_checker=True)
```
Comment (dalf) [inline searx/webapp.py:1353]: move the initialization at the end of the file, so the plugins can overwrite things.

**Comment by return42** [comment]
FYI I needed to fix initialize() aa5a5147b2  .. while I on the way I used a generator 5731b6b .. if you think generator fails for any reason, please drop 5731b6b.

I haven't finished my tests today, I will continue the next days ..

File: searx/plugins/__init__.py [hunk 2]
```diff
@@ -130,74 +46,200 @@ def sync_resource(base_path, resource_path, name, target_dir, plugin_dir):
...
             utime(resource_path, ns=(dep_stat.st_atime_ns, dep_stat.st_mtime_ns))
-        except:
-            logger.critical('failed to copy plugin resource {0} for plugin {1}'.format(file_name, name))
-            exit(3)
+        except IOError:
+            logger.critical(
+                "failed to copy plugin resource {0} for plugin {1}".format(
+                    file_name, name
+                )
+            )
```
Comment (return42) [inline searx/plugins/__init__.py:54]: Nit .. do we need such a lot of new-lines?  I often use ..

```python
            logger.critical(
                "failed to copy plugin resource {0} for plugin {1}".format(
                    file_name, name ))
```
.. even though I know this is not up to standard / but I think it is more readable.

**Review by return42 — APPROVED** [review]
FYI: I tested with:

- https://github.com/asciimoo/searx_external_plugin_example
- https://github.com/return42/tgwf-searx-plugins

Patch aa5a5147b is needed and 5731b6b70 is optional.

**Pull request merged**